src/cborparser.c

```diff
@@ -1459,6 +1459,137 @@ CborError cbor_value_map_find_value(const CborValue *map, const char *string, Cb
     return err;
 }
 
+/**
+ * Attempts to find the value in map \a map that corresponds to the integer key
+ * entry \a integer. If the iterator \a value does not point to a CBOR map, the
+ * behaviour is undefined, so checking with \ref cbor_value_get_type or \ref
+ * cbor_value_is_map is recommended.
+ *
+ * If the item is found, it is stored in \a result. If no item is found
+ * matching the key, then \a result will contain an element of type \ref
+ * CborInvalidType. Matching is performed using integer compare.
+ *
+ * This function has a time complexity of O(n) where n is the number of
+ * elements in the map to be searched. In addition, this function is has O(n)
+ * memory requirement based on the number of nested containers (maps or arrays)
+ * found as elements of this map.
+ *
+ * \sa cbor_value_is_valid(), cbor_value_advance()
+ */
+CborError cbor_value_map_find_value_by_int (const CborValue *map, int key, CborValue *element)
+{
+    CborError err;
+
+    cbor_assert (cbor_value_is_map(map));
+    err = cbor_value_enter_container(map, element);
+    if (err) {
+        goto error;
+    }
+
+    while (!cbor_value_at_end(element)) {
+        /* find the non-tag so we can compare */
+        err = cbor_value_skip_tag(element);
+        if (err) {
+            goto error;
+        }
+        if (cbor_value_is_integer(element)) {
+            int result;
+            err = cbor_value_get_int_checked (element, &result);
```

Comment:
Agree. I will change to get_int64_checked.

Reply:
That won't help for the value `1ULL << 63` in the map. Or you need to check and ignore the `CborErrorDataTooLarge` error from here.

Reply:
Then, I think I will change `int key` to `uint64_t key`. And only accept signed integer.

Reply:
That will then fail for negative values.

CBOR has a wider range than any C type. There's no way around this.

Reply:
I don't expect the negative value is used as key to get the value, at least in my use case, I did not see that.

Failure for negative value is an expected behavior. We can document the limitation clearly.

Reply:
Negative values are common. See the COSE RFC, for example:
```
  The algorithms defined in this document can be found in Table 5.

              +-------+-------+---------+------------------+
              | Name  | Value | Hash    | Description      |
              +-------+-------+---------+------------------+
              | ES256 | -7    | SHA-256 | ECDSA w/ SHA-256 |
              | ES384 | -35   | SHA-384 | ECDSA w/ SHA-384 |
              | ES512 | -36   | SHA-512 | ECDSA w/ SHA-512 |
              +-------+-------+---------+------------------+

                      Table 5: ECDSA Algorithm Values
```
https://datatracker.ietf.org/doc/html/rfc8152#section-8.1

Reply:
I know negative values are common.

But here I am talking the value used as key in map.
At least in [CoRim](https://datatracker.ietf.org/doc/draft-birkholz-rats-corim/) or [CoSwid](https://datatracker.ietf.org/doc/draft-ietf-sacm-coswid), I do not see such use case.

If we have to choose between int64_t and uint64_t, then I prefer uint64_t as key in map.

Other choice to define two APIs: find_value_by_int(), find_value_by_uint().

Reply:
Any of the three has the same problem because all the `_get_xxx_checked` function will give you a `CborErrorDataTooLarge` error for some numbers. You need to handle that for any and all of the three.

Reply:
> You need to handle that for any and all of the three.

I dont understand. Why I have to handle that?
This is invalid case. Why I cannot just return error to caller?

Reply:
Because it's not an error. For example, try to find the value for key 1 in the following map:
```
{ 9223372036854775808: "positive", -9223372036854775809: "negative", 1: "one" }
```
This is a valid map. Do you agree that if you passed this map to your function, it should return a `CborValue`  pointing to that "one" without error? Or do you think it should fail to find "one" and report an error instead?

Please write the unit test with this case. Please write unit tests containing all other types of keys before the value being sought, and unit tests where the keys are tagged too.

Reply:
Very good example. Thanks you @thiagomacieira. Now I understand your concern.

My first thought is that: this should be an error from C perspective.
But if CBOR allows it, then we probably need define a cbor_int_t, such as 
```
typedef struct {
  bool is_negative;
  typedef struct {
    uint64_t uint64;
    int64_t int64;
  } value;
} cbor_int_t;
```

I don't feel comfortable to treat `CborErrorDataTooLarge` as valid case randomly.

I will withdraw this PR and think more on that.

Reply:
I wish you'd continue. At a minimum, the array API is fine and can be included now -- as soon as you write a unit test (you can for it, that is.

I don't want such a integer type because I don't think it's useful. I don't think the use of numbers lesser than INT_MIN is common at all, but they are possible so we need to handle them. The only question is how to handle them. It's safe to handle `CborErrorDataTooLarge` after you've confirmed it's an integer, because that error can't happen otherwise. Or we add an internal function that simply acts as a comparator, duplicating some code from the get_xxx_checked functions.